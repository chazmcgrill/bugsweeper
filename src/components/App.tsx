import React, { useState, useEffect } from 'react';
import Controls from './Controls';
import Board from './Board';
import Footer from './Footer';
import { GAME_CONFIG } from '../utils/config';
import * as utils from '../utils';
import { createNewBoard } from '../utils/createNewBoard';

interface GameState {
    grid: GridTile[];
    flagCounter: number;
    endFlag: boolean | string;
}

interface AppProps {
    customBugIndexes?: number[];
}

const DEFAULT_GAME_STATE = {
    grid: [],
    flagCounter: 0,
    endFlag: false,
};

const toggleGridItemFlag = (grid: GridTile[], id: number) => grid
    .map(item => item.id === id ? { ...item, flagged: !item.flagged } : item);

const App = ({ customBugIndexes }: AppProps) => {
    const [gameState, setGameState] = useState<GameState>(DEFAULT_GAME_STATE);

    const tileClick = (id: number): void => {
        const { grid, endFlag, flagCounter } = gameState;
        const currentItem = grid[id];

        if (!endFlag) {
            const ids = utils.newTileStatus(currentItem, grid);
            const newGrid = grid.map((item, index) => ids.includes(index) ? { ...item, showing: true } : item);

            let newEndFlag = utils.endStateCheck(newGrid, flagCounter);

            if (currentItem.neighbours === 9) newEndFlag = "Oops Game Over!"

            setGameState({ ...gameState, grid: newGrid, endFlag: newEndFlag })
        }
    }

    const tileRightClick = (id: number, event: React.MouseEvent<HTMLDivElement, MouseEvent>): void => {
        event.preventDefault();

        if (!gameState.endFlag) {
            const grid = toggleGridItemFlag(gameState.grid, id);
            const flagCounter = grid.filter(n => n.flagged).length;
            const endFlag = utils.endStateCheck(grid, flagCounter);

            setGameState({ grid, endFlag, flagCounter })
        }
    }

    const handleReset = (): void => {
        const grid = createNewBoard();
        setGameState({ ...DEFAULT_GAME_STATE, grid });
    }

    useEffect(() => {
        const grid = createNewBoard(customBugIndexes);
        setGameState({ ...DEFAULT_GAME_STATE, grid });
    }, [customBugIndexes]);

    const { grid, endFlag, flagCounter } = gameState;

    return (
        <div className="app">
            <div className="container">
                <Controls
                    bugs={GAME_CONFIG.bugs}
                    endGame={endFlag}
                    flagCounter={flagCounter}
                />
                <Board
                    grid={grid}
                    handleTileClick={tileClick}
                    handleRightClick={tileRightClick}
                />
                <button onClick={handleReset}>New Game</button>
            </div>
            <Footer />
        </div>
    );
}

export default App;
