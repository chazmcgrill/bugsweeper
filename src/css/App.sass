.container
    font-family: 'Roboto Mono', monospace
    font-size: 22px
    margin: 40px auto
    width: 100%
    height: 430px
    border-radius: 4px
    display: flex
    align-items: center
    flex-direction: column
    justify-content: space-between

    @media screen and (min-width: 640px)
        font-size: 18px

.game-board
    width: 305px
    border: 5px solid $light
    padding: 5px
    text-align: center
    margin: 0 auto
    display: grid
    grid-template-columns: repeat(10, 1fr)
    grid-template-rows: repeat(10, 1fr)
    grid-gap: 5px
    border-radius: 5px

.tile
    background: $light
    border-radius: 2px
    height: 25px
    width: 25px
    cursor: pointer
    display: flex
    align-items: center
    justify-content: center

    &:hover
        background: $contrast
    
button
    font-family: inherit
    font-size: inherit
    padding: 0.3em 0.7em
    background: $bg
    border: 2px solid $dark
    border-radius: 5px
    cursor: pointer
    color: $dark
    transition: color 0.5s, background 0.5s
    outline: none
    z-index: 1

    &:hover
        color: $bg
        background: $dark

