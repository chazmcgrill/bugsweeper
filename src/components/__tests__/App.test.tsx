import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import App from '../App';

jest.mock('../../utils/config', () => ({
    get GAME_CONFIG() {
        return {
            width: 10,
            height: 10,
            bugs: 2
        }
    }
}))

test('shows initial message', () => {
    const { asFragment } = render(<App />);
    expect(screen.getAllByText('Find 2 bugs')).toBeTruthy();
    expect(screen.getAllByTestId('tile')).toHaveLength(100);
    expect(asFragment()).toMatchSnapshot();
});

test('right click tile adds flag icon', async () => {
    const { getByTestId, getAllByTestId } = render(<App />);        
    const eventNode = getAllByTestId('tile')[0];

    fireEvent.contextMenu(eventNode);

    const tile = await getByTestId('flag-icon');

    expect(tile).toBeTruthy();    
});

test('right click tile updates bug count', async () => {
    const { getByText, getAllByTestId } = render(<App customBugIndexes={[0]} />);
    const eventNode = getAllByTestId('tile')[0];

    fireEvent.contextMenu(eventNode);

    const titleText = await getByText('Find 1 bugs');

    expect(titleText).toBeTruthy();
});

test('pressing on bug ends game', async () => {
    const { getByText, getAllByTestId } = render(<App customBugIndexes={[0]} />);
    const eventNode = getAllByTestId('tile')[0];

    fireEvent.click(eventNode);

    const titleText = await getByText('Oops Game Over!');

    expect(titleText).toBeTruthy();
});

test('pressing on bug shows bug icon', async () => {
    const { getByTestId, getAllByTestId } = render(<App customBugIndexes={[0]} />);
    const eventNode = getAllByTestId('tile')[0];

    fireEvent.click(eventNode);

    const tile = await getByTestId('bug-icon');

    expect(tile).toBeTruthy();  
});

test('pressing on non bug clears tiles', async () => {
    const { getAllByTestId } = render(<App customBugIndexes={[0]} />);
    const eventNode = getAllByTestId('tile')[5];

    fireEvent.click(eventNode);

    const tile = await getAllByTestId('tile-showing');

    expect(tile).toBeTruthy(); 
});

test('pressing new game resets board', async () => {
    const { getByText, getAllByTestId } = render(<App customBugIndexes={[0]} />);
    const eventNode = getAllByTestId('tile')[5];

    fireEvent.click(eventNode);

    await getAllByTestId('tile-showing');

    const button = getByText('New Game');

    fireEvent.click(button);

    const tiles = await getAllByTestId('tile');

    expect(getByText('Find 2 bugs')).toBeTruthy();
    expect(tiles).toHaveLength(100); 
});

test('on flagging all bugs win state is shown', async () => {
    const { getByText, getAllByTestId } = render(<App customBugIndexes={[0, 1]} />);
    const bug1 = getAllByTestId('tile')[0];
    const bug2 = getAllByTestId('tile')[1];
    const nonBug = getAllByTestId('tile')[99];

    fireEvent.contextMenu(bug1);
    fireEvent.contextMenu(bug2);
    fireEvent.click(nonBug);

    const headerText = await getByText('You Win!');

    expect(headerText).toBeTruthy();
});