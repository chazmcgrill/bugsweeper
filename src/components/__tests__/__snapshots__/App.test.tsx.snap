// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows initial message 1`] = `
<DocumentFragment>
  <div
    class="app"
  >
    <div
      class="container"
    >
      <div
        class="controls"
      >
        <span>
          Find 2 bugs
        </span>
      </div>
      <div
        class="game-board"
      >
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
        <div
          class="tile"
          data-testid="tile"
        />
      </div>
      <button>
        New Game
      </button>
    </div>
    <footer>
      <div
        class="footer-icons"
      >
        <a
          href="https://www.freecodecamp.com/chazmcgrill"
        >
          <i
            aria-hidden="true"
            class="fa fa-free-code-camp"
          />
        </a>
        <a
          href="https://twitter.com/charlietcoder"
        >
          <i
            aria-hidden="true"
            class="fa fa-twitter"
          />
        </a>
        <a
          href="https://github.com/chazmcgrill"
        >
          <i
            aria-hidden="true"
            class="fa fa-github"
          />
        </a>
        <a
          href="https://www.linkedin.com/in/charliejbtaylor/"
        >
          <i
            aria-hidden="true"
            class="fa fa-linkedin"
          />
        </a>
        <a
          href="https://codepen.io/chazmcgrill/"
        >
          <i
            aria-hidden="true"
            class="fa fa-codepen"
          />
        </a>
      </div>
      <p
        class="footer-message"
      >
        designed and coded by
        <a
          href="http://www.charlietaylorcoder.com"
        >
           charlie taylor
        </a>
      </p>
    </footer>
  </div>
</DocumentFragment>
`;
